src/main/scala/scalaz/stream/async/TaskDriver.scala

```diff
@@ -0,0 +1,164 @@
+package scalaz.stream.async
+
+import scalaz.stream.Process
+import scalaz.stream.Process.{Halt, Step, Emit, Await, Cont}
+import scalaz.{Trampoline, \/}
+import scalaz.\/._
+import scalaz.concurrent.{Strategy, Actor, Future, Task}
+import scalaz.stream.Cause
+import scalaz.stream.Cause.{Kill, EarlyCause}
+import scalaz.stream.Util._
+
+/**
+ * Asynchronous driver for Process, where `F` is specialized to TASK 
+ */
+object TaskDriver {
+
+  /**
+   * Runs the supplied task with possible chance to interrupt it.
+   * Task is interrupted when resulting callback is applied.
+   *
+   * Whenever task is interrupted, following is guaranteed:
+   * - if, there was asynchronous code of task in process, the `cb` will get invoked immediatelly with `Task.TaskInterrupted` exception
+   * - if the task is complete, then interruption is no-op
+   * - if the task is running `pure` code, that code is completed and `cb` is invoked with result of that code
+   *
+   * Whenever task is completed, following is guaranteed
+   * - if the interrupt was received while evaluating the task (pure code) then `cb` is called with result of the task evaluation
+   * - if the interrupt was received while evaluating the task (async code) then evaluated result is discarded
+   * - otherwise `cb` is called with result of task evaluatin
+   *
+   * @param task    Task to evaluate
+   * @param cb      cb to invoke after evalutation is done, or when task is interrupted
+   */
+  def runTaskAsyncInterruptibly[A](task: Task[A], cb: (Throwable \/ A) => Unit): () => Unit = {
+    sealed trait M
+    case object Interrupt extends M
+    case class Complete(r: Throwable \/ A) extends M
+    case class CompleteBind(r: Any, g: Any => Future[Throwable \/ A]) extends M
+
+    var open: Boolean = true
+
+    def complete(r: (Throwable \/ A)): Unit = {
+      open = false
+      cb(r)
+    }
+
+    def tryCompleteTask(current: Future[Throwable \/ A], actor: => Actor[M]): Unit = {
+      current.step match {
+        case Future.Now(r) => complete(r)
+        case Future.Async(onFinish) =>
+          onFinish((r: Throwable \/ A) => Trampoline.done(actor ! Complete(r)))
+        case Future.BindAsync(onFinish, g) =>
+          onFinish((r: Any) => Trampoline.done(actor ! CompleteBind(r, g)))
+
+      }
+    }
+
+    lazy val actor: Actor[M] = Actor[M]({
+      case Complete(r) if open => complete(r)
+      case CompleteBind(r, g) if open => tryCompleteTask(g(r), actor)
+      case Interrupt if open => complete(left(Task.TaskInterrupted))
+      case _ => //no-op
```

Comment:
This actually defines `onComplete` finalizers with an "at most once" semantic, not "exactly once".  If a task is interrupted in an `Async` case (not `BindAsync`), that value will be lost entirely and any associated resources will leak.  I have a test which catches this in #326.  True "exactly once" semantics are very difficult to achieve with `runAsync` (I think impossible, actually), but the implementation in #326 approaches this more closely with "exactly once" semantics for non-await-dependent `onComplete`s and "at most once" semantics for await-dependent `onComplete`s with the added guarantee that await-dependent `onComplete`s will be run provided their awaited task produces a value (this added guarantee is what is missing here).

Put another way, this case means that `tryCompleteTask` is actually no different from `runAsyncInterruptibly`!  If this is in fact the semantic that you want, then you may as well use the standard interpreter.